cas-mfa-java/src/main/java/net/unicon/cas/mfa/authentication/principal/ServiceSpecificPrincipalAttributeMultiFactorAuthenticationRequestResolver.java

```diff
@@ -0,0 +1,258 @@
+        Pattern pattern = translationMap.get(attributePattern);
+        if (pattern == null) {
+            pattern = Pattern.compile(attributePattern);
+        }
+
+        return pattern.matcher(attributeValue).matches();
+    }
+
+    /**
+     * Looks up the mfa data for a specific service.
+     * @param targetService the service to check
+     * @return service specific mfa settings
+     */
+    private ServiceMfaData getServicesAuthenticationData(final WebApplicationService targetService) {
+        final RegisteredService registeredService = this.servicesManager.findServiceBy(targetService);
+        if (registeredService == null) {
+            logger.debug("No registered service is found. Delegating to the next argument extractor in the chain...");
+            return null;
+        }
+
+        if (!(registeredService instanceof RegisteredServiceWithAttributes)) {
+            logger.debug("Registered service is not capable of defining an mfa attribute.");
+            return null;
+        }
+
+        final ServiceMfaData serviceData = new ServiceMfaData();
+
+        final RegisteredServiceWithAttributes service = RegisteredServiceWithAttributes.class.cast(registeredService);
+
+        final Map mfaRole = Map.class.cast(service.getExtraAttributes().get("mfa_role"));
+        if (mfaRole == null) {
+            return null;
+        }
+
+        serviceData.setAttributeName(String.class.cast(mfaRole.get("mfa_attribute_name")));
+        serviceData.setAttributePattern(String.class.cast(mfaRole.get("mfa_attribute_pattern")));
+        serviceData.setAuthenticationMethod(String.class.cast(service.getExtraAttributes().get("authn_method")));
+
+        return serviceData;
+    }
+
+    /**
+     * A POJO to store data retrieved from the Services Manager.
+     */
+    private class ServiceMfaData {
+        private String authenticationMethod;
+        private String attributeName;
+        private String attributePattern;
+
+        public String getAuthenticationMethod() {
+            return authenticationMethod;
+        }
+
+        public void setAuthenticationMethod(final String authenticationMethod) {
+            this.authenticationMethod = authenticationMethod;
+        }
+
+        public String getAttributeName() {
+            return attributeName;
+        }
+
+        public void setAttributeName(final String attributeName) {
+            this.attributeName = attributeName;
+        }
+
+        public String getAttributePattern() {
+            return attributePattern;
+        }
+
+        public void setAttributePattern(final String attributePattern) {
+            this.attributePattern = attributePattern;
+        }
+
+        /**
+         * Checks the the class stores valid data.
+         * @return true if valid data, otherwise false
+         */
+        public boolean isValid() {
+            if (this.authenticationMethod != null && this.attributeName != null && this.attributePattern != null) {
```

Comment:
Break this apart please, with distinct log messages for each condition. 

Also use isNotBlank().